agent/cache/cache.go

```diff
@@ -836,11 +872,46 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 			r.Info.MinIndex = 0
 			c.fetch(key, r, false, attempt, true)
 		}
-	}()
+	}(handle)
 
 	return entry.Waiter
 }
 
+func (c *Cache) getOrReplaceFetchHandle(key string) fetchHandle {
+	c.fetchLock.Lock()
+	defer c.fetchLock.Unlock()
+
+	if prevHandle, ok := c.fetchHandles[key]; ok {
+		close(prevHandle.stopCh)
+	}
+
+	c.lastFetchID++
```

Comment:
Why do we need to manage an ID at the cache level? I think it will be sufficient to use a unique id per key. Which mean that we can use the last `fetchHandles` id and increment it.
If we do so we could possibly use a sync Map for this.

Reply:
> I think it will be sufficient to use a unique id per key

This is to identity specific goroutines for the _same_ key which is the exact bug here. Does my explanation above help explain why it's needed?

Or maybe I'm misundestanding.

> we could possibly use a sync Map for this

In general `sync.Map` is almost never the right choice - they are specifically designed for cases like reloadable config where reads are extremely common and updates are so rare that it's effectively never. This case will be writing the value every minutes and only marginally less than it reads I'd guess which means a mutex is almost certainly the better approach unless profiling tells us otherwise!

From the docs:

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

Reply:
Thank you both for explaining this. Til that sync.Map is almost never a good choice 😓